smartapps/bangali/rooms-child-app.src/rooms-child-app.groovy

```diff
@@ -1241,18 +1241,21 @@ private pageRoomTemperature() {
             input "maintainRoomTemp", "enum", title: "Maintain room temperature?", required: false, multiple: false, defaultValue: 4,
                                                                             options: [[1:"Cool"], [2:"Heat"], [3:"Both"], [4:"Neither"]], submitOnChange: true
             if (['1', '2', '3'].contains(maintainRoomTemp))     {
+				if (personsPresence)
```

Comment:
Just housekeeping and thermostatOffset, which is the "ensure trigger" offset.

Reply:
trying to understand thermostatOffset since i dont use a central thermostat. does setting the thermostat to a lower or higher value then issuing a thermostat.cool() or thermostat.heat() not turn on the thermostat?

thanks.

Reply:
To simplify, let's assume all sensors read the same at the same actual temperature or it's not possible to discern one as they seem to be all over the place, that's a separate concern. Let's also assume we want the area around the thermostat and the room at the same temperature, 70F. Finally, the overall assumption is, when we are heating this is the minimum temperature we want, we don't mind being a little warmer but certainly not colder.

Here's the scenario: The thermostat is at 70F and not heating  but the room is at 68F.  We want to heat the room +2F. With no sensor offset and no thermostat offset, we would make the setpoint 70F, but that would do nothing.

With a thermostat offset of, say, 4F, we would make it 74F instead which would ensure triggering the thermostat. When the room reaches 70F, we turn off the thermostat by setting it to the current thermostat temperature (which would presumably be ~72F)

Reply:
ok. do we really need the thermostat offset then? more settings usually tend to confuse users. why dont we use a static offset of 5 degrees and update the text on that page to note that.

also, we keep talking about this as a central thermostat but the handling seems to be on a per room basis. heres an example:
room A maintain: 68F room sensor: 65F temperature offset: 2F
room B maintain: 68F room sensor: 72F temperature offset: 5F
room C maintain: 68F room sensor: 70F temperature offset: 3F

this will keep toggling the thermostat ... that wont be good.

thanks.

Reply:
Oh yeah, it doesn't really make sense on a per room basis. Sure, hard coded 5F sounds fine, since we don't really have a suitable spot for it in the parent app.

Reply:
hmm, thanks.

how about the 2nd part ... the frequent toggle on off problem?

Reply:
Re: the last paragraph - the design of the parent app doesn't help with something like this, where the main interface just links to the child apps and everything else is stuck under Next

Reply:
Here's the only more centralized model I can think of. It'll probably work better:

1) Leave the thermostat option in, but remove the actual thermostat device selection from room settings. 
2) Get rid of the trigger offset (keep sensor deltas)
3) Add the central thermostat(s) selection to room manager settings.
4) When a rule is active and the central thermostat is picked, do nothing except set requestHeat/Cool in room state
5) Room manager runs a method periodically to check each room's state, calculating an average temperature of all rooms asking for HVAC.
6) It then reads setpoint and temperature differential from the central thermostat, and adjusts setpoint so the differential between the calculated average temperature and the new setpoint is identical. The setpoint is verified every single time so when it's changed beyond our control, a new differential is calculated.

Reply:
i am researching multi-zone thermostats a bit. will update tomorrow.

Reply:
From Ecobee: https://support.ecobee.com/hc/en-us/articles/115004907227-How-do-room-sensors-work-

Ecobee changes its temperature reading, we can't do that so we have to simulate it. Reading what I wrote above for#6.... that's not quite right, this is simpler and better:

1-If we are notManaging and the thermostat has been active within the past X minutes (10? Adjustable?), abort
2-If we are Managing, check to see if the setpoint was changed outside our control. If so abort to notManaging state, send notification so they can realize they are fighting the automation and don't do anything for a timeout period (1 hr? Adjustable?). 
3- Otherwise re-evaluate the setpoint based on the difference between room sensor readings active rules and set it.

On my end I'll whip up something and put it into testing (although it's warming up here :/ ) but let me know if you see anything immediately wrong with it...

Reply:
from looking around seems like there is really no true multi-zone systems. multi-zone systems seems to be more like multiple systems that handle different zones. there is some pseudo multi-zone system using active vents but not something that is worth addressing for now.

with that in mind think the approach in processCoolHeat method should be:
1) find all rooms with the same thermostat
2) find active temperature rule for each room based on rules match
3) calculate averages of heat to, cool to and temperature range from matched rules
4) use the calculated averages to manage thermostat heating, cooling and set point

thoughts?

Reply:
So what you are suggesting takes over the thermostat, instead of just taking into account room temperatures and "nudging" it. That's going to be more effective obviously - I've already written the "nudging" code but haven't really seen it in action yet, it just got cold again here.

Sounds good to me... There should be a warning if common thermostats are found.. something like "A shared thermostat has been detected. Please cancel any built-in or smart app programming for this thermostat, and ensure enough temperature sensors available around the house to effectively manage it."

Reply:
Also thermostat and on/off room heating shouldn't be exclusive I guess. One could have central HVAC and room heater/AC.

Reply:
right. since with any typical thermostat there is really no way to just heat or cool a room. might as well look at all rooms with that thermostat, find the averages of the rules and temperatures and use these average numbers to manage the thermostat.

yeah its quite likely folks use thermostat along with in room heaters may be even some room acs. this is where it complexity goes up by an order at least.

Reply:
#1 & #2 will require quite a bit of change.

lets start with #1 first? may be if you can write up a bit on how you want to handle it between various rooms and how data from them is aggregated and evaluated prior to managing the thermostat commands? that way you dont code it all up and me come along with - how about we do it that way instead? :-)

Reply:
Sure. Start a new PR?

Reply:
sure.
1) close this PR
2) any changes from this PR that are unrelated to thermostat related changes move to new PR and I can review and merge that
3) open a new PR with thermostat related changes to handle this centralized concept and we can discuss changes there.

Reply:
havent heard from you for a while ... :-)

Reply:
Busy with moving :) The plan sounds good. I'll probably get started on Sunday. I may submit a couple of other PRs first actually (adjacency & auto level)  they are already running on my system but gotta clean up the commits...

Reply:
ahh ... that can be consuming but hopefully not stressful. :-)

👍

Reply:
guessing you are still tied up with the after move effects? :-) do you want me to get the thermostat related changes started? i also have an user asking for fan support for some time now so want to get that done as well for him. but was holding off till the thermostat changes were final.

sorry this is not meant to bug you.

thanks.

Reply:
ok. let me start with #1 for thermostat. 

> right. since with any typical thermostat there is really no way to just heat or cool a room. might as well look at all rooms with that thermostat, find the averages of the rules and temperatures and use these average numbers to manage the thermostat.

one thing i am debating about this is whether i should transparently override user intent ... generally i am not for doing that. instead of finding all the rules and averaging them out would it better if and when they add the same thermostat more than one time to different rooms warn them that `this may have undesirable effect ... blah ... blah ...` and not transparently override it in the background.

any thoughts on this?

thanks.

Reply:
A notification is needed but it can be subtle: "Shared thermostat, temperature averaging enabled"
They can figure it out from there.